monix-execution/shared/src/main/scala/monix/execution/misc/CanBindLocals.scala

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2014-2019 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.execution.misc
+
+import java.util.concurrent.CompletableFuture
+import java.util.function.BiFunction
+import monix.execution.{CancelableFuture, FutureUtils}
+import monix.execution.schedulers.TrampolineExecutionContext
+import scala.concurrent.Future
+
+/**
+  * Type class for describing how isolation work for specific data types.
+  * There is a default instance for any type but some of them (e.g. `Future`) require
+  * special handling and using type class is an alternative to overloads and `asInstanceOf` calls.
+  */
+trait CanBindLocals[R] {
+  def withSuspendedContext(ctx: () => Local.Context)(f: => R): R
+
+  def withContext(ctx: Local.Context)(f: => R): R =
+    withSuspendedContext(() => ctx)(f)
+
+  def bind[A](local: Local[A], value: Option[A])(f: => R): R =
+    withSuspendedContext(() => Local.getContext().bind(local.key, value))(f)
+
+  /** Execute a  block of code without propagating any `Local.Context`
+    * changes outside.
+    */
+  def isolate(f: => R): R =
+    withSuspendedContext(CanBindLocals.mkIsolatedRef)(f)
+}
+
+object CanBindLocals extends CanIsolateInstancesLevel1 {
+  def apply[R](implicit R: CanBindLocals[R]): CanBindLocals[R] = R
+
+  private val mkIsolatedRef =
+    () => Local.getContext().isolate()
+}
+
+private[misc] abstract class CanIsolateInstancesLevel1 extends CanIsolateInstancesLevel0 {
+  /**
+    * Instance for `monix.execution.CancelableFuture`.
+    */
+  implicit def cancelableFuture[R]: CanBindLocals[CancelableFuture[R]] =
+    FutureInstance.asInstanceOf[CanBindLocals[CancelableFuture[R]]]
+
+  /**
+    * Instance for `java.util.concurrent.CompletableFuture`.
+    */
+  implicit def completableFuture[R]: CanBindLocals[CompletableFuture[R]] =
+    CompletableFutureInstance.asInstanceOf[CanBindLocals[CompletableFuture[R]]]
```

Comment:
Can we do this on 2.11? It's not guaranteed to have Java 8, which is when CompletableFuture was added.

Reply:
Ah, good point. You're right, it's only available starting with Java 8.